### ISSUE TITLE
Question: Roles of nodes in memberlist

### ISSUE THREAD

**Issue by diptanu** [hashicorp/memberlist#46]
Hi,

Do you think it would be possible to designate role or additional metadata regarding a node when it joins a cluster? I understand that it's currently not possible but would be interested in a PR which would allow a user to configure the role of a node and query memberlist for nodes with specific metadata?

We are currently using the name field and adding more information to the name and parse the names to decipher the roles of various nodes in a cluster.

**Comment by armon** [comment]
@diptanu You can! You want to use the `NodeMeta` to associate metadata with the node (https://github.com/hashicorp/memberlist/blob/master/delegate.go#L10). We use this in Serf to support Key/Value tags. The queries can be done locally since the nodes have the full member list with all the meta data. In most cases, you are better off using the Serf library instead of Memberlist directly as it implements a ton of both nice and necessary (partition recovery) features.

**Comment by armon** [comment]
Serf is confusingly both a CLI based agent, but also just a drop in library. Consul embeds the Serf library for example for all gossip related features.

**Comment by diptanu** [comment]
@armon Thanks a ton! This library is awesome :) Also, are you suggesting to use serf as a library instead of memberlist?

Also, I need to put raft on top of serf like you guys do in Consul, however I need to do multiple smaller concensus groups within a cluster instead of a single concensus group.

**Comment by armon** [comment]
@diptanu Yeah the Serf library is both easier to use and has a ton more features. This library was meant to be almost purely an implementation of the SWIM algorithm, but many of the bells and whistles needed for the real world are in Serf. Sounds like you are working on something interesting, I'd love to learn more :)

**Comment by kyleterry** [comment]
Given you recommend using Serf as a library, are there any docs on doing that? I've been looking at the Consul code to see how this is done, and that gets me pretty far, but it's a lot of context switching.

Thanks for both of these things!

**Comment by JensRantil** [comment]
@kyleterry How about opening an issue with the Serf project/repo instead?

@kyleterry @armon I suggest this issue can be closed.

**Comment by longquanzheng** [comment]
> @diptanu You can! You want to use the `NodeMeta` to associate metadata with the node (https://github.com/hashicorp/memberlist/blob/master/delegate.go#L10). We use this in Serf to support Key/Value tags. The queries can be done locally since the nodes have the full member list with all the meta data. In most cases, you are better off using the Serf library instead of Memberlist directly as it implements a ton of both nice and necessary (partition recovery) features.

Just curious, why does serf depend on memberlist?
